### PR TITLE
Binary quantization

### PR DESCRIPTION
**Pull request by irevoire**
## Related issue
Fixes https://github.com/meilisearch/arroy/issues/69

## What does this PR do?
- Introduce a new `UnalignedVector` generic type to replace the `UnalignedF32Slice` type we were using before
  - It's parametrized by a `Codec`
  - Supports the unaligned `f32` (equivalent to the old `UnalignedF32Slice` type)
  - Supports the binary quantized slice by converting each f32 to a single `-1` or `1` depending on if their value was negative or positive
  - It can convert the binary quantized slice to `Vec<f32>` quickly using SIMD
- Provide new distance trait to binary quantized the `Euclidean`, `Manhattan` and `Angular` distances, respectively named `BinaryQuantizedEuclidean`, `BinaryQuantizedManhattan` and `BinaryQuantizedAngular`
- To keep the relevancy good while binary quantizing, we had to re-develop `two_means` « un-binary quantize » the vectors before searching for the centroid

----

> [!WARNING]
> Below is the initial investigation of the different method we could use to make the binary quantization works without losing too much relevancy.

## First version

Just binary quantize every operation.

Here's the measured relevancy:
image

## Second version to improve relevancy

One issue we found out is that when binary quantizing vectors, we basically end up creating a bunch of clusters in two dimensions. It would look like this:
image

All vectors will end up on one of the four edges of the square.

The more dimensions we have, the more clusters we'll get.
The number of clusters is `2^nb_dimensions`.

By making the internal computation of `two_means` **not** use the binary quantized vectors but instead use the real vectors, here are the results:

image

> [!WARNING]
> It’s actually worse than initially 

## Third idea to improve relevancy

In the second solution I was computing the `two_means` loop with **non** binary-quantized distances which greatly improved the relevancy.
But then the output of `two_means` was binary quantized again.
We should try to compute the `normal` on **non** binary quantized distance as well and then binary quantize this vector right before storing it in the DB.

image

> [!NOTE]  
> This improved the relevancy by almost 10 points of recall in the worst case over the previous best solution.

<details><summary>With the bits being [0:1] the relevancy is terrible</summary>
image
</details>

## Fourth solution to improve relevancy

Store the **non** binary quantized distance in the `SplitNode` in the database directly.
This increase by a lot the size of the database (still way less than the non-binary-quantized distance though).
The search may become slower as well.

The results are bad and I can't explain why: 
image

I made another branch in case we want to investigate further: https://github.com/meilisearch/arroy/pull/84

----

## In conclusion

The best version is the third one.

The next steps are:
- Optimize all the binary quantized version
- Merge + make a release
- Merge in meilisearch + find a way to change the distances => We'll talk about that with louis in two weeks after our vacations

-------

## Next step in Arroy:

1. Add the size of the DB in the relevancy benchmarks
2. Overfetch search results (between x3 and x6)
2. Compare ourselves to qdrant
3. Optimize performances

### CODE REVIEW

File: src/unaligned_vector/binary_quantized.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            .chunks_exact(size_of::<QuantizedWord>())
+            .map(|bytes| QuantizedWord::from_ne_bytes(bytes.try_into().unwrap()))
+            .flat_map(|mut word| {
+                let mut ret = vec![0.0; QUANTIZED_WORD_SIZE];
+                for index in 0..QUANTIZED_WORD_SIZE {
+                    let bit = word & 1;
+                    word >>= 1;
+                    if bit == 1 {
+                        ret[index] = 1.0;
+                    }
```
Comment (Kerollmops) [inline src/unaligned_vector/binary_quantized.rs:55]:
```suggestion
                    ret[index] = bit as f32;
```

File: src/unaligned_vector/binary_quantized.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    fn from_vec(vec: Vec<f32>) -> Cow<'static, UnalignedVector<Self>> {
+        Cow::Owned(Self::from_slice(&vec).into_owned())
+    }
+
+    fn iter(vec: &UnalignedVector<Self>) -> impl Iterator<Item = f32> + '_ {
+        vec.vector
+            .chunks_exact(size_of::<QuantizedWord>())
+            .map(|bytes| QuantizedWord::from_ne_bytes(bytes.try_into().unwrap()))
+            .flat_map(|mut word| {
+                let mut ret = vec![0.0; QUANTIZED_WORD_SIZE];
```
Comment (Kerollmops) [inline src/unaligned_vector/binary_quantized.rs:49]:
```suggestion
                let mut ret = [0.0; QUANTIZED_WORD_SIZE];
```

File: src/unaligned_vector/binary_quantized.rs (hunk 1 above)
Comment (Kerollmops) [inline src/unaligned_vector/binary_quantized.rs:55]: Can be optimized to avoid storing 512 bits in memory

File: src/distance/binary_quantized_euclidean.rs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    fn margin_no_header(
+        p: &UnalignedVector<Self::VectorFormat>,
+        q: &UnalignedVector<Self::VectorFormat>,
+    ) -> f32 {
+        dot_product(p, q)
+    }
+}
+
+fn dot_product(u: &UnalignedVector<BinaryQuantized>, v: &UnalignedVector<BinaryQuantized>) -> f32 {
+    u.as_bytes().iter().zip(v.as_bytes()).map(|(u, v)| (u ^ v).count_ones()).sum::<u32>() as f32
```
Comment (Kerollmops) [inline src/distance/binary_quantized_euclidean.rs:78]: Could you put a comment that relevancy can be impacted here? But for non-module 64 dimensions.

File: src/writer.rs [hunk 4]
```diff
@@ -87,7 +87,11 @@ impl<D: Distance> Writer<D> {
 
     /// Returns an `Option`al vector previous stored in this database.
     pub fn item_vector(&self, rtxn: &RoTxn, item: ItemId) -> Result<Option<Vec<f32>>> {
-        Ok(item_leaf(self.database, self.index, rtxn, item)?.map(|leaf| leaf.vector.into_owned()))
+        Ok(item_leaf(self.database, self.index, rtxn, item)?.map(|leaf| {
+            let mut vec = leaf.vector.to_vec();
+            vec.drain(self.dimensions..);
```
Comment (Kerollmops) [inline src/writer.rs:92]:
```suggestion
            vec.set_len(self.dimensions);
```

File: src/unaligned_vector/mod.rs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+    /// Returns an iterator of f32 that are read from the vector.
+    /// The f32 are copied in memory and are therefore, aligned.
+    pub fn iter(&self) -> impl Iterator<Item = f32> + '_ {
+        Codec::iter(self)
+    }
+
+    /// Returns an allocated and aligned `Vec<f32>`.
+    pub fn to_vec(&self) -> Vec<f32> {
+        self.iter().collect()
```
Comment (Kerollmops) [inline src/unaligned_vector/mod.rs:79]: use capacity with len here

File: src/node.rs [hunk 6]
```diff
@@ -184,11 +112,26 @@ impl fmt::Debug for ItemIds<'_> {
...
     pub left: NodeId,
     pub right: NodeId,
-    pub normal: Cow<'a, UnalignedF32Slice>,
+    pub normal: Cow<'a, UnalignedVector<D::VectorCodec>>,
+}
+
+impl<D: Distance> fmt::Debug for SplitPlaneNormal<'_, D> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("SplitPlaneNormal")
+            .field("left", &self.left)
```
Comment (Kerollmops) [inline src/node.rs:124]: Can you add the `D::name` under a `distance_name` field?

**Review by Kerollmops — CHANGES_REQUESTED** [review]

File: src/unaligned_vector/binary_quantized.rs [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+                let offset = output_ptr.add(current_byte * 8 + i * 4);
+                vst1q_f32(offset, lane);
+            }
+        }
+    }
+
+    output
+}
+
+#[cfg(not(any(target_arch = "aarch64", target_arch = "arm64ec")))]
```
Comment (Kerollmops) [inline src/unaligned_vector/binary_quantized.rs:192]: This is not the right feature clause. Use the right one. Same for the neon one.

File: src/unaligned_vector/binary_quantized.rs [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+    current_element: QuantizedWord,
+    current_iteration: usize,
+    iter: ChunksExact<'a, u8>,
+}
+
+impl Iterator for BinaryQuantizedIterator<'_> {
+    type Item = f32;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current_iteration >= QUANTIZED_WORD_SIZE {
```
Comment (Kerollmops) [inline src/unaligned_vector/binary_quantized.rs:272]: Rename it something like `QUANTIZED_WORD_BITS`

File: examples/relevancy.rs
Comment (Kerollmops) [inline examples/relevancy.rs:1]: Remove this test but add a link to the vector test repository somewhere in the README (make it pretty).

**Review by Kerollmops — CHANGES_REQUESTED** [review]

File: src/node.rs (hunk 6 above)
Comment (irevoire) [reply src/node.rs:125]: Done in https://github.com/meilisearch/arroy/pull/82/commits/af598ca76e900dba23b0966e8078eeb40b0e9da2

File: src/unaligned_vector/binary_quantized.rs (hunk 8 above)
Comment (irevoire) [reply src/unaligned_vector/binary_quantized.rs]: Done in https://github.com/meilisearch/arroy/pull/82/commits/71f6659bb4aeed3ea64d536ba566034f66c0ad4d

File: src/unaligned_vector/binary_quantized.rs (hunk 7 above)
Comment (irevoire) [reply src/unaligned_vector/binary_quantized.rs]: Fixed in https://github.com/meilisearch/arroy/pull/82/commits/777effb409fc0ab9f797f93ba8eded676b6bb0c8

File: examples/relevancy.rs
Comment (irevoire) [reply examples/relevancy.rs:1]: Fixed in https://github.com/meilisearch/arroy/pull/82/commits/393a5c578baee2de70f3a1e62a1f4794b302c3a2

**Review by Kerollmops — APPROVED** [review]
It looks perfect to me 👌 That is a wonderful job that you've done here @irevoire 👏
Thank you!
